### DISCUSSION TITLE
Widget Ideas - What integrations would you like to see added?

### DISCUSSION THREAD

**Discussion by lissy93** [Lissy93/dashy#370] [Ideas]
As requested, am working on adding support for dynamic content, in the form of widgets
While testing I've created a few (system stats, Pi-Hole, cron monitoring, uptime, weather, RSS, etc). But I want to hear your ideas - What widgets and integrations would you like to see added?

I'm hoping to get this feature finished and merged by Christmas, so stay tuned!

:)

[![sneak previws](https://i.ibb.co/JsxzXxj/81868f25-892e-490b-a374-20d250e62c5c.png)](https://github.com/Lissy93/dashy/)

**Comment by EVOTk** [comment]
Hello,
It's just beautiful!
Personally here is what could interest me:

- Uptime (uptime-kuma integration? Https://github.com/louislam/uptime-kuma)
- Pi-Hole / Adguard
- Plex (Tautulli) / Jellyfin (streaming count?)
- rutorrent (Active Torrent, Download and Upload Speed) and all that revolves around (Sonarr, Radarr, Prowlarr, ...)

**Reply by lissy93** [reply]
Great ideas, thank you! :)

That should all be possible, and fun to build, so I will get started on those now :)

Actually was working on some Pi-Hole integrations this weekend, here's a sneak preview:
Though with AdGuard, I don't AdGuard has an API, so that one might not be possible :(

<p align="center"><img width="300" src="https://i.ibb.co/zftCLJN/pi-hole-stats.png" /></p>
<p align="center"><img width="300" src="https://i.ibb.co/7kdxxwx/pi-hole-recent-queries.png" /></p>

**Reply by EVOTk** [reply]
It's really pretty :) I can't wait to be able to set this up!
I have 2 Pi-Hole instances, will it be possible to deploy the same widget twice but for 2 different instances?

Do you plan to make different widgets for the same service?
For example for PiHole, a full widget, like your example, and a more simplistic one, for example with only the status and the statistics circles?

**Reply by lissy93** [reply]
Yup, that's possible, widgets work like items, so you can have 2 or more of the same widget but with different settings.
And yeah, the screenshot above is actually 2 widgets, the pi-hole-traffic-chart and the pi-hole-stats. Here is the (draft) [widget docs](https://github.com/Lissy93/dashy/blob/FEATURE/basic-widget-support/docs/widgets.md#self-hosted-services-widgets) I have been working on

**Reply by EVOTk** [reply]
good work ! Well done to you and thank you again for everything you do!

Uptime Kuma does not have an API but it is requested! This could be a good idea for Dashy widgets! I hope the API will one day be integrated into uptime kuma! https://github.com/louislam/uptime-kuma/issues/118

**Reply by lissy93** [reply]
I love Uptime Kuma, the best status monitoring app I've used, so I hope it gets an API endpoint in the future. I noticed they it has a [Prometheus Endpoint](https://github.com/louislam/uptime-kuma/wiki/Prometheus-Integration), so it might be possible to use that data, I will try it out...

**Comment by EVOTk** [comment]
Hello,
Is a translation of the Widgets planned? :)

**Reply by lissy93** [reply]
Where possible I will make the text translatable, but a lot of the data displayed is directly from the API response, so unless that service supports multiple languages, then some of the results might not be possible to translate.

- Wherever an API supports the users current language, it will be applied
- For dates and times, they will be automatically displayed in your language
- Any hard-coded text can be translated, and I will add it to the translations file :)

**Reply by EVOTk** [reply]
> Where possible I will make the text translatable, but a lot of the data displayed is directly from the API response, so unless that service supports multiple languages, then some of the results might not be possible to translate.

Of course !

Great :) I can't wait to try this new feature :)

**Comment by stuu3k** [comment]
Hey, few ideas....

-SabNZBd+ would certainly be useful with an option to start, stop and pause and a view of 1-5 downloads for example or even just a status summary.
-GMAIL \ OtherMail Unread Counts
-Network UPS Plugin, view battery states and levels etc.  (https://networkupstools.org/)

Another thing which would be nice would be the ability to control home automation tasks, though maybe you can do this with an API URL and a shortcut anyway not sure, but a button to turn my lights on without speaking to Google \ Alexa \ Jarvis whatever would be amazing.

Last one probably a crap load of work though and maybe a bit out of scope.

Thanks

**Reply by lissy93** [reply]
Up next I'm working on sub-items and multi-page support (see [roadmap](https://github.com/Lissy93/dashy/discussions/405) for more info), and the next widgets I'm building are some multi-media widgets, Overseerr, Plex etc.

I can look into your suggestions, but it probably won't be for a couple months.
For future reference, if you could provide the API endpoints + sample output of the API, then there will be a higher chance that will be implemented. There's a section in the [widget docs](https://github.com/Lissy93/dashy/blob/master/docs/widgets.md#requesting-a-widget) about requesting widgets. Alternatively, [here's a guide](https://github.com/Lissy93/dashy/blob/master/docs/development-guides.md#building-a-widget) on building a widget yourself.

Sorry, for any disappointment, this thread is quite old and left over from before there were any widgets built.

**Reply by lissy93** [reply]
Just had a look at your list...
- Controlling home automation tasks will be able to be done through Action buttons, which is on the roadmap :)
- SabNZBd should be possible, but I'm not familiar with their API. Might come back to it in the future
- Don't know what that brand of UPS is, but it doesn't appear to be open source or have an API, so that will be a no
- I'm not keen on the Gmail one, Google APIs are a pain to work with, but possibly a generic LDAP, CalDav plugin? Though you'd need an intermediary service to create a REST API to expose the data.

**Reply by stuu3k** [reply]
Just FYI on the UPS thing, actually most do use a standard they just don't advertise that fact very well, I would suggest trying the opensource app "UPSMon" which is also on Github and seeing if that can connect, if you can get that working there is no reason why you cannot do it.

**Comment by lissy93** [comment]
This thread is now out-dated. For up-to-date info about requesting a widget, please see [widget docs](https://github.com/Lissy93/dashy/blob/master/docs/widgets.md#requesting-a-widget).
Now that everything's in place, developing your own widget is quite straightforward, and is outlined in [this guide](https://github.com/Lissy93/dashy/blob/master/docs/development-guides.md#building-a-widget).

**Reply by patrickheeney** [reply]
@Lissy93 I was looking to develop some local only widgets to interface with some private APIs but not run a custom fork of dashy. I read the guide but it looks like it needs to be [registered](https://github.com/Lissy93/dashy/blob/master/docs/development-guides.md#step-3---register) in the component. Would it be possible to have widgets read from the yaml file at runtime so new widgets could be added. In docker we would mount the code from the filesystem ~/dashy/code:/code and then have something like this for the config. Might be able to load them via [http](https://github.com/FranckFreiburger/vue3-sfc-loader) as well.

```
dynamicwidgets:
  - name: ExampleWidget
    source: /code/Widgets/ExampleWidget.vue
```

**Reply by lissy93** [reply]
Yeah, well kind of.
You could just build a generic web component, then either load it as an [iframe widget ](https://github.com/Lissy93/dashy/blob/master/docs/widgets.md#iframe-widget) or an [embed widget](https://github.com/Lissy93/dashy/blob/master/docs/widgets.md#html-embedded-widget).

**Reply by patrickheeney** [reply]
@Lissy93 Is there any interested in that as a feature request? Or do you prefer to only have internal widgets and use iframe or embed as for external. I know you have an upcoming api widget as well. That might work if user can supply html with some template or variable substitution mapped to the api response.

**Reply by lissy93** [reply]
>  Is there any interested in that as a feature request? 

Sorry, not sure I understand?

If your widget is for an open source, managed service or anything that may be useful to others, then submitting a PR back upstream, would both be easiest for you, and make it available to others. We've already had several widgets contributed by the community, and I hope there will be more to come :)

**Reply by patrickheeney** [reply]
Sorry for the typos: ```Is there any interest in that as a feature request?```. Specifically in being able to load in widgets at runtime that are user defined and created that do not exist in the open source package. Or even checking a user mapped directory and not having to define them at all.

In this case it is a private web service api for a company so it is not something public or a publicly available service. While an iframe or embed could be developed it would be great to write a custom vue components that leverage the company API that already exists rather than try and have the company provide a iframe or embedded widget.

**Reply by lissy93** [reply]
Ah sorry, I understand now.
So that was actually how I initially planned for it to work, since I didn't want to include 50+ widget components in the main repo when most users will only use 1 or 2. But I don't think it's technically possible, or at least not neatly. As the widget components need to be available at build time. If you know a way of achieving this neatly, I would very much be open to it.


So I think your best option would be to build a generic JS or iframe widget. That would also mean your not tied into Dashy, as you could easily use it elsewhere if in the future you wanted to switch to a different dashboard app.

Alternatively, creating a private fork of the project, and using a GH action to keep it synced with upstream Dashy, and publish the image to a private Docker registry.

**Reply by lissy93** [reply]
Your changes are awesome, nice one

**Reply by patrickheeney** [reply]
@Lissy93 Thank you. I haven't done any testing yet but my initial hope was to mount a volume and have vue pick it up by looking in the right directory. I am not sure if that mounted volume would work with the original vue files or if they would need pre-compiled. I wrote some fake config of what I was thinking in [this PR](https://github.com/Lissy93/dashy/pull/758). 

One other idea is to use something like https://github.com/FranckFreiburger/vue3-sfc-loader. It looks like it is possible to pull in single file vue components. I am not sure if they could be pulled in over http or if this could pull them in from the filesystem. 

Either one is not too clean but ideally dashy would bundle a good set of core widgets and more niche or dependency heavy ones would be included separately somehow. Maybe the lite build has the core ones but the full build has everything if they can't be bundled separately.

**Reply by lissy93** [reply]
That first method, would actually work.
I've put a working example is in #761.
It lets you mount a directory of components, or single component (annoyingly I think this would need to be in the src directory), then list that / those components in your config, which would update your COMPT array and import it.

It's still not the neatest, and I do really want to have a more dynamic option, as that would also allow me to also move a lot of the existing components into a separate repo to lighten things up a bit.
